experimental/sql/src/projection_pushdown.py

```diff
@@ -66,6 +74,42 @@ def find_schema(self, op: RelAlg.Operator) -> list[str]:
       return [s.data for s in op.names.data]
     if isinstance(op, RelAlg.Aggregate):
       return [s.data for s in op.col_names.data]
+    if isinstance(op, RelAlg.CartesianProduct):
+      return self.find_schema(op.left.op) + self.find_schema(op.right.op)
+    if isinstance(op, RelAlg.OrderBy):
+      return self.find_schema(op.input.op)
+    if isinstance(op, RelAlg.Limit):
+      return self.find_schema(op.input.op)
+
+
+@dataclass
+class ProjectionSimplifier(ProjectionOptimizer):
+
+  # This rewriter simpifies projections such that they only project columns that
+  # are acutlaly needed upstream.
+
+  @op_type_rewrite_pattern
+  def match_and_rewrite(self, op: RelAlg.Project, rewriter: PatternRewriter):
+    cols = list(dict.fromkeys(self.find_cols_upstream(op.parent_op())))
+    if None in cols or len(op.names.data) == len(cols):
+      return
+
+    input_schema = self.find_schema(op)
+    output_schema = self.flatten(
+        [self.find_cols_in_expr(o) for o in op.projections.ops])
+
+    if input_schema != output_schema or [s.data for s in op.names.data
```

Comment:
What does this test do? If the input is *not* different from the output, they are the same, so the projection didn't do anything, no? Is this the case that is handled here? If so, I don't understand it from the documentation/comments...

Reply:
I added more comments to clarify.